app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -164,60 +155,55 @@ class SubscriptionController @Inject()
-                validDesAddress => {
-                  mark("Count-Subscription-AddressLookup-Success")
-                  subscribe(details, validDesAddress).map(redirectSubscriptionResponse)
-                }
-              )
-            }
-          }.getOrElse(Future.successful(sessionMissingRedirect()))
-        }
+  def returnFromAddressLookup(id: String): Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
+      sessionStoreService.fetchInitialDetails.flatMap { maybeDetails =>
+        maybeDetails.map { details =>
+          addressLookUpConnector.getAddressDetails(id).flatMap { address =>
+            desAddressForm.bindAddressLookupFrontendAddress(details.utr, address).fold(
+              formWithErrors => Future successful Ok(html.address_form_with_errors(formWithErrors)),
+              validDesAddress => {
+                mark("Count-Subscription-AddressLookup-Success")
+                subscribe(details, validDesAddress).map(redirectSubscriptionResponse)
+              })
+          }
+        }.getOrElse(Future.successful(sessionMissingRedirect()))
+      }
+    }
   }
 
-  def submitModifiedAddress: Action[AnyContent] = AuthorisedWithSubscribingAgentAsync() {
-    implicit authContext =>
-      implicit request =>
-        desAddressForm.form.bindFromRequest().fold(
-          formWithErrors => Future successful Ok(html.address_form_with_errors(formWithErrors)),
-          validDesAddress =>
-            sessionStoreService.fetchInitialDetails.flatMap { maybeInitialDetails =>
-              maybeInitialDetails.map { initialDetails =>
-                subscribe(initialDetails, validDesAddress).map(redirectSubscriptionResponse)
-              }.getOrElse(Future.successful(sessionMissingRedirect()))
-            }
-        )
+  def submitModifiedAddress: Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
+      desAddressForm.form.bindFromRequest().fold(
+        formWithErrors => Future successful Ok(html.address_form_with_errors(formWithErrors)),
+        validDesAddress =>
+          sessionStoreService.fetchInitialDetails.flatMap { maybeInitialDetails =>
+            maybeInitialDetails.map { initialDetails =>
+              subscribe(initialDetails, validDesAddress).map(redirectSubscriptionResponse)
+            }.getOrElse(Future.successful(sessionMissingRedirect()))
+          })
+    }
   }
 
-  val showSubscriptionComplete: Action[AnyContent] = AuthorisedWithAgentAffinityAsync {
-    implicit authContext =>
-      implicit request => {
-        val arn = request.flash.get("arn")
-
-        arn match {
-          case Some(arn) =>
-            sessionStoreService.fetchContinueUrl.
-              recover { case NonFatal(ex) =>
+  val showSubscriptionComplete: Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
```

Comment:
Need to be careful here due to APB-2118 - "Subscription complete page is skipped (sometimes)". 

There's often a latency before auth is updated to reflect the new enrolment in the QA environment, but in Live it's often visible from auth immediately.

There's a redirect in the `AuthActions.withSubscribingAgent` to the AS Account Homepage if the enrolment is seen. This redirect would skip the complete page.
So.. we'd probably just need to secure this page with a simple  `AuthActions.withAuthenticatedUser`

Reply:
The corresponding test for this situation is in `SubscriptionControllerISpec`'s "tolerste a possible short delay in the new enrolment becoming visible in auth"...